### PR TITLE
Type hinting developer docs

### PR DESCRIPTION
**Pull request by ksunden**
## PR summary

- Update devdocs with more details about type hinting, especially deprecations
- Bring existing parameter keyword only deprecations in line with documentation
  - Deprecated and slated for removal deprecations are not added, though going forward we suggest not removing the hints where they already exist until expiry.

I considered adding something regarding the behavior of the alias decorator (namely that the aliases are not type hinted, added to the `mypy-stubtest-allowlist.txt` file)
But there was not already a section about that decorator, and it is rare enough that it gets _added_ that I don't necessarily think its warrented to write one just for this.

## PR checklist

- [N/A] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [N/A] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-examples-and-tutorials)
- [N/A] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-docstrings) guidelines

### CODE REVIEW

File: doc/devel/contribute.rst [hunk 1]
```diff
@@ -345,6 +345,8 @@ Introducing
    - Use ``_api.warn_deprecated()`` for general deprecation warnings
    - Use the decorator ``@_api.deprecated`` to deprecate classes, functions,
      methods, or properties
+   - Use the decorator ``@_api.deprecate_privatize_attribute`` to deprecate
```
Comment (QuLogic) [inline doc/devel/contribute.rst:348]: This isn't a decorator; it's a fancy descriptor thing.

File: doc/devel/contribute.rst [hunk 2]
```diff
@@ -354,6 +356,23 @@ Introducing
...
+    introduction.
+  - Items decorated with ``@_api.rename_parameter`` or ``@_api.make_keyword_only``
+    report the *new* (post deprecation) signature at runtime, and thus *should* be
+    updated on introduction.
+  - Items decorated with ``@_api.delete_parameter`` should include a default value hint
+    for the deleted parameter, even if it did not previously have one (e.g.
+    ``param: <type> = ...``). Even so, the decorator changes the default value to a
+    sentinel value which should not be included in the type stub. Thus, Mypy Stubtest
+    needs to be informed of the inconsistency by placing the method into
+    ``ci/mypy-stubtest-allowlist.txt`` under a heading indicating the deprecation
```
Comment (QuLogic) [inline doc/devel/contribute.rst:373]:
```suggestion
    :file:`ci/mypy-stubtest-allowlist.txt` under a heading indicating the deprecation
```

File: doc/devel/contribute.rst [hunk 3]
```diff
@@ -365,6 +384,19 @@ Expiring
    information. For the content, you can usually copy the deprecation notice
    and adapt it slightly.
 2. Change the code functionality and remove any related deprecation warnings.
+3. Make appropriate changes to the type hints in the associated ``.pyi`` file.
+
+  - Items marked with ``@_api.deprecated`` or ``@_api.deprecate_privatize_attribute``
+    are to be removed on expiry.
+  - Items decorated with ``@_api.rename_parameter`` or ``@_api.make_keyword_only``
+    will have been updated at introduction.
```
Comment (QuLogic) [inline doc/devel/contribute.rst:392]:
```suggestion
    will have been updated at introduction, and require no change now.
```

File: doc/devel/contribute.rst [hunk 4]
```diff
@@ -365,6 +384,19 @@ Expiring
...
+  - Items decorated with ``@_api.delete_parameter`` will need to be updated to the
+    final signature, in the same way as the ``.py`` file signature is updated.
+    The entry in ``ci/mypy-stubtest-allowlist.txt`` should be removed.
+  - Any other entries in ``ci/mypy-stubtest-allowlist.txt`` under a versions
```
Comment (QuLogic) [inline doc/devel/contribute.rst:396]:
```suggestion
    The entry in :file:`ci/mypy-stubtest-allowlist.txt` should be removed.
  - Any other entries in :file:`ci/mypy-stubtest-allowlist.txt` under a version's
```

**Review by QuLogic — APPROVED** [review]

**Review by timhoffm — APPROVED** [review]

**Pull request merged**